src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java

```diff
@@ -732,16 +733,19 @@ public NavigableSet<Long> listCyclesBetween(int lowerCycle, int upperCycle) thro
     public <T> void addCloseListener(Closeable key) {
         synchronized (closers) {
             if (!closers.isEmpty())
-                closers.removeIf(Closeable::isClosed);
-            closers.add(key);
+                closers.removeIf(wrc -> {
+                    final Closeable closeable = wrc.get();
+                    return (closeable != null) ? closeable.isClosed() : true;
```

Comment:
We don't need to keep empty weak references around do we? 
`return closeable == null || closeable.isClosed()`

Reply:
Sorry that's what it does, but perhaps refactor to less byzantine logic